go/vt/vttablet/tabletserver/health_streamer.go

```diff
@@ -0,0 +1,182 @@
+	degradedThreshold  time.Duration
+	unhealthyThreshold time.Duration
+
+	mu      sync.Mutex
+	clients map[chan *querypb.StreamHealthResponse]struct{}
+	state   *querypb.StreamHealthResponse
+
+	history *history.History
+}
+
+func newHealthStreamer(env tabletenv.Env, alias topodatapb.TabletAlias) *healthStreamer {
+	return &healthStreamer{
+		stats:              env.Stats(),
+		degradedThreshold:  env.Config().Healthcheck.DegradedThresholdSeconds.Get(),
+		unhealthyThreshold: env.Config().Healthcheck.UnhealthyThresholdSeconds.Get(),
+		clients:            make(map[chan *querypb.StreamHealthResponse]struct{}),
+
+		state: &querypb.StreamHealthResponse{
+			Target:      &querypb.Target{},
+			TabletAlias: &alias,
+			RealtimeStats: &querypb.RealtimeStats{
+				HealthError: errUnintialized,
+			},
+		},
+
+		history: history.New(5),
+	}
+}
+
+func (hs *healthStreamer) InitDBConfig(target querypb.Target) {
+	hs.state.Target = &target
+}
+
+func (hs *healthStreamer) Stream(ctx context.Context, callback func(*querypb.StreamHealthResponse) error) error {
+	ch := hs.register()
+	defer hs.unregister(ch)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case shr := <-ch:
+			if err := callback(shr); err != nil {
+				if err == io.EOF {
+					return nil
+				}
+				return err
+			}
+		}
+	}
+}
+
+func (hs *healthStreamer) register() chan *querypb.StreamHealthResponse {
+	hs.mu.Lock()
+	defer hs.mu.Unlock()
+
+	ch := make(chan *querypb.StreamHealthResponse, 1)
+	hs.clients[ch] = struct{}{}
+
+	// Send the current state immediately.
+	ch <- proto.Clone(hs.state).(*querypb.StreamHealthResponse)
+	return ch
+}
+
+func (hs *healthStreamer) unregister(ch chan *querypb.StreamHealthResponse) {
+	hs.mu.Lock()
+	defer hs.mu.Unlock()
+
+	delete(hs.clients, ch)
+}
+
+func (hs *healthStreamer) ChangeState(tabletType topodatapb.TabletType, terTimestamp time.Time, lag time.Duration, err error, serving bool) {
+	hs.mu.Lock()
+	defer hs.mu.Unlock()
+
+	hs.state.Target.TabletType = tabletType
+	if tabletType == topodatapb.TabletType_MASTER {
+		hs.state.TabletExternallyReparentedTimestamp = terTimestamp.Unix()
+	} else {
+		hs.state.TabletExternallyReparentedTimestamp = 0
+	}
+	if err != nil {
+		hs.state.RealtimeStats.HealthError = err.Error()
+	} else {
+		hs.state.RealtimeStats.HealthError = ""
+	}
+	hs.state.RealtimeStats.SecondsBehindMaster = uint32(lag.Seconds())
+	hs.state.Serving = serving
+
+	hs.state.RealtimeStats.SecondsBehindMasterFilteredReplication, hs.state.RealtimeStats.BinlogPlayersCount = blpFunc()
+	hs.state.RealtimeStats.Qps = hs.stats.QPSRates.TotalRate()
+
+	shr := proto.Clone(hs.state).(*querypb.StreamHealthResponse)
+
+	for ch := range hs.clients {
+		select {
+		case ch <- shr:
+		default:
+		}
+	}
+	hs.history.Add(&historyRecord{
+		Time:       time.Now(),
+		serving:    shr.Serving,
+		tabletType: shr.Target.TabletType,
+		lag:        lag,
+		err:        err,
+	})
+}
+
+func (hs *healthStreamer) ApppendDetails(details []*kv) []*kv {
```

Comment:
I do not see any test for this method

Reply:
There's no framework to test the `/debug/status` pages right now. But I tested it manually though.